src/core/cgroup.c

```diff
@@ -2422,6 +2422,29 @@ void unit_release_cgroup(Unit *u) {
         }
 }
 
+int unit_maybe_release_cgroup(Unit *u) {
+        int r;
+
+        assert(u);
+
+        if (!u->cgroup_path)
+                return 1;
+
+        /* Don't release the cgroup if there are still processes under it. If we get notified later when all the
+         * processes exit (e.g. the processes were in D-state and exited after the unit was marked as failed)
+         * we need the cgroup paths to continue to be tracked by the manager so they can be looked up and cleaned
+         * up later. */
+        r = cg_is_empty_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path);
+        if (r < 0)
+                return log_unit_debug_errno(u, r, "Error checking if the cgroup is recursively empty, ignoring: %m");
+        if (r == 0)
+                log_unit_debug(u, "Not releasing %s because it is still populated", u->cgroup_path);
+        else
+                unit_release_cgroup(u);
+
+        return r;
+}
```

Comment:
this functin should probably return 1 if it actually did something, and 0 if not.

That way callers will know if anything was released